src/reflect/scala/reflect/internal/Constants.scala

```diff
@@ -243,28 +264,18 @@ trait Constants extends api.Constants {
     def typeValue: Type     = value.asInstanceOf[Type]
     def symbolValue: Symbol = value.asInstanceOf[Symbol]
 
-    /**
-     * Consider two `NaN`s to be identical, despite non-equality
-     * Consider -0d to be distinct from 0d, despite equality
-     *
-     * We use the raw versions (i.e. `floatToRawIntBits` rather than `floatToIntBits`)
-     * to avoid treating different encodings of `NaN` as the same constant.
-     * You probably can't express different `NaN` varieties as compile time
-     * constants in regular Scala code, but it is conceivable that you could
-     * conjure them with a macro.
-     */
-    private def equalHashValue: Any = value match {
-      case f: Float  => floatToRawIntBits(f)
-      case d: Double => doubleToRawLongBits(d)
-      case v         => v
-    }
-
     override def hashCode: Int = {
       import scala.util.hashing.MurmurHash3._
       val seed = 17
       var h = seed
       h = mix(h, tag.##) // include tag in the hash, otherwise 0, 0d, 0L, 0f collide.
-      h = mix(h, equalHashValue.##)
+      val valueHash = tag match {
+        case NullTag => 0
+        case FloatTag => java.lang.Float.hashCode(floatToRawIntBits(value.asInstanceOf[Float]))
+        case DoubleTag => java.lang.Double.hashCode(doubleToRawLongBits(value.asInstanceOf[Double]))
```

Comment:
I believe should just be
```scala
java.lang.Double.hashCodevalue.asInstanceOf[Double])
```
Currently you're converting the `Long` returned by `doubleToRawLongBits` back to a `Double` to give it to `hashCode`.

Same for the `Float` case.

Reply:
Thanks. On review, I meant to write `java.lang.{Integer, Long}.hashCode({float,double}ToRaw...Bits(value))`.

-0f and 0f must not considered equal, which is why `equals` uses the `rawBits`. I suppose we don't need to go through `rawBits` to have a consistent hash code (even though that's the most obviously correct implementation.) I think `value.hashCode` would be consistent with equals (although those zeros would collide)

Reply:
I've pushed a fix with the non-colliding version with a comment that we could alternatively just use `value.hashCode`.

Reply:
The zero's won't collide:
```scala
scala> java.lang.Double.hashCode(0d)
res0: Int = 0

scala> java.lang.Double.hashCode(-0d)
res1: Int = -2147483648
```
All the NaNs will collide, though:
```scala
scala> java.lang.Double.hashCode(java.lang.Double.NaN)
res3: Int = 2146959360

scala> java.lang.Double.doubleToRawLongBits(java.lang.Double.NaN)
res4: Long = 9221120237041090560

scala> java.lang.Double.longBitsToDouble(res4+1)
res5: Double = NaN

scala> java.lang.Double.hashCode(res5)
res6: Int = 2146959360

scala> res3 == res6
res7: Boolean = true
```